timemachine/fe/mcgregor.py

```diff
@@ -135,45 +126,105 @@ def __init__(self, n_vertices, edges):
+        components. It is assumed that source is a mapped node. i.e. node_states[source] == NODE_STATE_VISITED_MAPPED.
+        If this function returns True, then the resulting graph is definitely disconnected. If this function returns
+        False, then the resulting graph can still be connected.
+
+        For example, let M be mapped node, D be a demapped node, and U be an unvisited node. If
+        `max_connected_components=1`:
 
         1. M-U-M # returns False
         2. M-D-M # returns True
         3. M-M-U # returns False
         4. M-M-D # returns False
 
-        The implementation is adapted from the _plain_bfs() method in:
-        https://networkx.org/documentation/stable/_modules/networkx/algorithms/components/connected.html
-
+        This calls an optimized implementation in mapping_is_disconnected_fast; see mapping_is_disconnected_ref for a
+        simpler-to-understand reference version.
         """
-        unseen = [True] * self.n_vertices
-        unseen[source] = False
-        nextlevel = [source]
-        mapped_count = 1
-
-        if mapped_count == n_mapped_nodes:
-            return False
-
-        while nextlevel:
-            thislevel = nextlevel
-            nextlevel = []
-            for v in thislevel:
-                for w in self.get_neighbors(v):
-                    if unseen[w] and node_states[w] != NODE_STATE_VISITED_DEMAPPED:
-                        if node_states[w] == NODE_STATE_VISITED_MAPPED:
-                            mapped_count += 1
-                            if mapped_count == n_mapped_nodes:
-                                return False
-                        unseen[w] = False
-                        nextlevel.append(w)
-        return True
+
+        # Uncomment to assert result of fast implementation matches the reference
+        # args = (mapped_nodes, unvisited_nodes, max_connected_components, min_connected_component_size)
+        # assert self.mapping_is_disconnected_fast(*args) == self.mapping_is_disconnected_ref(*args)
+
+        return self.mapping_is_disconnected_fast(
+            mapped_nodes, unvisited_nodes, max_connected_components, min_connected_component_size
+        )
+
+    def mapping_is_disconnected_fast(
+        self,
+        mapped_nodes: Set[int],
+        unvisited_nodes: Set[int],
+        max_connected_components: Optional[int],
+        min_connected_component_size: int,
+    ):
+        """Optimized implementation of mapping_is_disconnected, sacrificing modularity for efficiency"""
+
+        seen = set()
+        n_ccs = 0
+        for v in mapped_nodes:
+            if v not in seen:
+                # visit component containing v
+                seen.add(v)
+                cc_size = 1
+                nextlevel = [v]
+                while nextlevel:
+                    thislevel = nextlevel
+                    nextlevel = []
+                    for v in thislevel:
```

Comment:
nit: shadowing the original v value, this is safe as written but could bite someone later

Reply:
Renamed to avoid shadowing in https://github.com/proteneer/timemachine/pull/1349/commits/c3d4bcbacf81018878d00c54d07eaf94e5493c1f